File: C#/WP7/source/LFX-0.3.3/Loon.Core.Graphics.Component/LMessage.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using Loon.Action.Sprite;
using Loon.Core.Graphics.OpenGL;
using Loon.Core.Input;

namespace Loon.Core.Graphics.Component
{
    public class LMessage : LContainer
    {
        private Animation animation;

        private LFont messageFont = LFont.GetDefaultFont();

        private LColor fontColor = LColor.white;

        private long printTime, totalDuration;

        private int dx, dy, dw, dh;

        private Print print;

        public LMessage(int width, int height): this(0, 0, width, height)
        {
           
        }

        public LMessage(int x, int y, int width, int height):this((LTexture)null, x, y, width, height)
        {
            
        }

        public LMessage(String fileName, int x, int y):this(LTextures.LoadTexture(fileName), x, y)
        {
            
        }

        public LMessage(LTexture formImage, int x, int y):this(formImage, x, y, formImage.GetWidth(), formImage.GetHeight())
        {
            
        }

        public LMessage(LTexture formImage, int x, int y, int width, int height): base(x, y, width, height)
        {
            this.animation = new Animation();
            if (formImage == null)
            {
                this.SetBackground(new LTexture(width, height, true));
                this.SetAlpha(0.3F);
            }
            else
            {
                this.SetBackground(formImage);
                if (width == -1)
                {
                    width = formImage.GetWidth();
                }
                if (height == -1)
                {
                    height = formImage.GetHeight();
                }
            }
            this.print = new Print(GetLocation(), messageFont, width, height);
            if (XNAConfig.IsActive())
            {
                this.SetTipIcon(XNAConfig.LoadTex(LSystem.FRAMEWORK_IMG_NAME + "creese.png"));
            }
            this.totalDuration = 80;
            this.customRendering = true;
            this.SetWait(false);
            this.SetElastic(true);
            this.SetLocked(true);
            this.SetLayer(100);
        }

        public void SetWait(bool flag)
        {
            print.SetWait(flag);
        }

        public bool IsWait()
        {
            return print.IsWait();
        }

        public void Complete()
        {
            print.Complete();
        }

        public void SetLeftOffset(int left)
        {
            print.SetLeftOffset(left);
        }

        public void SetTopOffset(int top)
        {
            print.SetTopOffset(top);
        }

        public int GetLeftOffset()
        {
            return print.GetLeftOffset();
        }

        public int GetTopOffset()
        {
            return print.GetTopOffset();
        }

        public int GetMessageLength()
        {
            return print.GetMessageLength();
        }

        public void SetMessageLength(int messageLength)
        {
            print.SetMessageLength(messageLength);
        }

        public void SetTipIcon(String fileName)
        {
            print.SetCreeseIcon(new LTexture(fileName));
        }

        public void SetTipIcon(LTexture icon)
        {
            print.SetCreeseIcon(icon);
        }

        public void SetNotTipIcon()
        {
            print.SetCreeseIcon(null);
        }

        public void SetEnglish(bool e)
        {
            print.SetEnglish(true);
        }

        public bool IsEnglish()
        {
            return print.IsEnglish();
        }

        public void SetDelay(long delay)
        {
            this.totalDuration = ((delay < 1) ? (long)(1) : (long)(delay));
        }

        public long GetDelay()
        {
            return totalDuration;
        }

        public bool IsComplete()
        {
            return print.IsComplete();
        }

        public void SetPauseIconAnimationLocation(int dx_0, int dy_1)
        {
            this.dx = dx_0;
            this.dy = dy_1;
        }

        public void SetMessage(String context, bool isComplete)
        {
            print.SetMessage(context, messageFont, isComplete);
        }

        public void SetMessage(String context)
        {
            print.SetMessage(context, messageFont);
        }

        public String GetMessage()
        {
            return print.GetMessage();
        }

        /// <summary>
        /// 处理点击事件（请重载实现）
        /// </summary>
        ///
        public virtual void DoClick()
        {
            if (Click != null)
            {
                Click.DownClick(this, input.GetTouchX(), input.GetTouchY());
                Click.UpClick(this, input.GetTouchX(), input.GetTouchY());
            }
        }

        protected internal override void ProcessTouchClicked()
        {
            this.DoClick();
        }

        protected internal override void ProcessKeyPressed()
        {
            if (this.IsSelected() && this.input.GetKeyPressed() == Key.ENTER)
            {
                this.DoClick();
            }
        }

        public override void Update(long elapsedTime)
        {
            if (!visible)
            {
                return;
            }
            base.Update(elapsedTime);
            if (print.IsComplete())
            {
                animation.Update(elapsedTime);
            }
            printTime += elapsedTime;
            if (printTime >= totalDuration)
            {
                printTime = printTime % totalDuration;
                print.Next();
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        protected override void CreateCustomUI(GLEx g, int x, int y, int w,
                int h)
        {
            if (!visible)
            {
                return;
            }
            LFont oldFont = g.GetFont();
            g.SetFont(messageFont);
            print.Draw(g, fontColor);
            g.SetFont(oldFont);
            if (print.IsComplete() && animation != null)
            {
                if (animation.GetSpriteImage() != null)
                {
                    g.SetAlpha(1.0F);
                    UpdateIcon();
                    g.DrawTexture(animation.GetSpriteImage(), dx, dy);
                }
            }
            g.ResetColor();
        }

        protected internal override void ProcessTouchDragged()
        {
            if (!locked)
            {
                if (GetContainer() != null)
                {
                    GetContainer().SendToFront(this);
                }
                this.Move(this.input.GetTouchDX(), this.input.GetTouchDY());
                this.UpdateIcon();
            }
        }

        public void SetPauseIconAnimation(Animation a)
        {
            this.animation = a;
            if (a != null)
            {
                LTexture image = a.GetSpriteImage(0);
                if (image != null)
                {
                    this.dw = image.GetWidth();
                    this.dh = image.GetHeight();
                    this.UpdateIcon();
                }
            }
        }

        private void UpdateIcon()
        {
            this.SetPauseIconAnimationLocation(GetScreenX() + GetWidth() - dw / 2
                    - 20, GetScreenY() + GetHeight() - dh - 10);
        }

        public LColor GetFontColor()
        {
            return fontColor;
        }

        public void SetFontColor(LColor f)
        {
            this.fontColor = f;
        }

        public LFont GetMessageFont()
        {
            return messageFont;
        }

        public void SetMessageFont(LFont f)
        {
            this.messageFont = f;
        }

        public bool IsLocked()
        {
            return locked;
        }

        public void SetLocked(bool locked)
        {
            this.locked = locked;
        }

        protected internal override void ValidateSize()
        {
            base.ValidateSize();
        }

        public override void CreateUI(GLEx g, int x, int y, LComponent component,
                LTexture[] buttonImage)
        {

        }

        public override String GetUIName()
        {
            return "Message";
        }

        public override void Dispose()
        {
            base.Dispose();
            if (print != null)
            {
                print.Dispose();
                print = null;
            }
            if (animation != null)
            {
                animation.Dispose();
                animation = null;
            }
        }
    }
}

File: Java/Android/OpenGLES/source/core/src/loon/core/graphics/component/LMessage.java
package loon.core.graphics.component;

import loon.action.sprite.Animation;
import loon.core.LSystem;
import loon.core.graphics.LColor;
import loon.core.graphics.LComponent;
import loon.core.graphics.LContainer;
import loon.core.graphics.LFont;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.graphics.opengl.LTextures;
import loon.core.graphics.opengl.LTexture.Format;
import loon.core.input.LInputFactory.Key;

/**
 * Copyright 2008 - 2009
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
public class LMessage extends LContainer {

	private Animation animation;

	private LFont messageFont = LFont.getDefaultFont();

	private LColor fontColor = LColor.white;

	private long printTime, totalDuration;

	private int dx, dy, dw, dh;

	private Print print;

	public LMessage(int width, int height) {
		this(0, 0, width, height);
	}

	public LMessage(int x, int y, int width, int height) {
		this((LTexture) null, x, y, width, height);
	}

	public LMessage(String fileName, int x, int y) {
		this(LTextures.loadTexture(fileName), x, y);
	}

	public LMessage(LTexture formImage, int x, int y) {
		this(formImage, x, y, formImage.getWidth(), formImage.getHeight());
	}

	public LMessage(LTexture formImage, int x, int y, int width, int height) {
		super(x, y, width, height);
		this.animation = new Animation();
		if (formImage == null) {
			this.setBackground(new LTexture(width, height, true, Format.SPEED));
			this.setAlpha(0.3F);
		} else {
			this.setBackground(formImage);
			if (width == -1) {
				width = formImage.getWidth();
			}
			if (height == -1) {
				height = formImage.getHeight();
			}
		}
		this.print = new Print(getLocation(), messageFont, width, height);
		this.setTipIcon(LSystem.FRAMEWORK_IMG_NAME + "creese.png");
		this.totalDuration = 80;
		this.customRendering = true;
		this.setWait(false);
		this.setElastic(true);
		this.setLocked(true);
		this.setLayer(100);
	}

	public void setWait(boolean flag) {
		print.setWait(flag);
	}

	public boolean isWait() {
		return print.isWait();
	}

	public void complete() {
		print.complete();
	}

	public void setLeftOffset(int left) {
		print.setLeftOffset(left);
	}

	public void setTopOffset(int top) {
		print.setTopOffset(top);
	}

	public int getLeftOffset() {
		return print.getLeftOffset();
	}

	public int getTopOffset() {
		return print.getTopOffset();
	}

	public int getMessageLength() {
		return print.getMessageLength();
	}

	public void setMessageLength(int messageLength) {
		print.setMessageLength(messageLength);
	}

	public void setTipIcon(String fileName) {
		print.setCreeseIcon(new LTexture(fileName));
	}

	public void setTipIcon(LTexture icon) {
		print.setCreeseIcon(icon);
	}

	public void setNotTipIcon() {
		print.setCreeseIcon(null);
	}

	public void setEnglish(boolean e) {
		print.setEnglish(true);
	}

	public boolean isEnglish() {
		return print.isEnglish();
	}

	public void setDelay(long delay) {
		this.totalDuration = (delay < 1 ? 1 : delay);
	}

	public long getDelay() {
		return totalDuration;
	}

	public boolean isComplete() {
		return print.isComplete();
	}

	public void setPauseIconAnimationLocation(int dx, int dy) {
		this.dx = dx;
		this.dy = dy;
	}

	public void setMessage(String context, boolean isComplete) {
		print.setMessage(context, messageFont, isComplete);
	}

	public void setMessage(String context) {
		print.setMessage(context, messageFont);
	}

	public String getMessage() {
		return print.getMessage();
	}

	/**
	 * 处理点击事件（请重载实现）
	 * 
	 */
	public void doClick() {
		if (Click != null) {
			Click.DoClick(this);
		}
	}

	@Override
	protected void processTouchClicked() {
		this.doClick();
	}

	@Override
	protected void processKeyPressed() {
		if (this.isSelected() && this.input.getKeyPressed() == Key.ENTER) {
			this.doClick();
		}
	}

	@Override
	protected void processTouchPressed() {
		if (!input.isMoving()) {
			if (Click != null) {
				Click.DownClick(this, input.getTouchX(), input.getTouchY());
			}
		}
	}

	@Override
	protected void processTouchReleased() {
		if (!input.isMoving()) {
			if (Click != null) {
				Click.UpClick(this, input.getTouchX(), input.getTouchY());
			}
		}
	}

	@Override
	public void update(long elapsedTime) {
		if (!visible) {
			return;
		}
		super.update(elapsedTime);
		if (print.isComplete()) {
			animation.update(elapsedTime);
		}
		printTime += elapsedTime;
		if (printTime >= totalDuration) {
			printTime = printTime % totalDuration;
			print.next();
		}
	}

	@Override
	protected synchronized void createCustomUI(GLEx g, int x, int y, int w,
			int h) {
		if (!visible) {
			return;
		}
		LFont oldFont = g.getFont();
		g.setFont(messageFont);
		print.draw(g, fontColor);
		g.setFont(oldFont);
		if (print.isComplete() && animation != null) {
			if (animation.getSpriteImage() != null) {
				g.setAlpha(1.0F);
				updateIcon();
				g.drawTexture(animation.getSpriteImage(), dx, dy);
			}
		}
		g.resetColor();
	}

	@Override
	protected void processTouchDragged() {
		if (!locked) {
			if (getContainer() != null) {
				getContainer().sendToFront(this);
			}
			this.move(this.input.getTouchDX(), this.input.getTouchDY());
			if (Click != null) {
				Click.DragClick(this, input.getTouchX(), input.getTouchY());
			}
			this.updateIcon();
		}
	}

	public void setPauseIconAnimation(Animation animation) {
		this.animation = animation;
		if (animation != null) {
			LTexture image = animation.getSpriteImage(0);
			if (image != null) {
				this.dw = image.getWidth();
				this.dh = image.getHeight();
				this.updateIcon();
			}
		}
	}

	private void updateIcon() {
		this.setPauseIconAnimationLocation(getScreenX() + getWidth() - dw / 2
				- 20, getScreenY() + getHeight() - dh - 10);
	}

	public LColor getFontColor() {
		return fontColor;
	}

	public void setFontColor(LColor fontColor) {
		this.fontColor = fontColor;
	}

	public LFont getMessageFont() {
		return messageFont;
	}

	public void setMessageFont(LFont messageFont) {
		this.messageFont = messageFont;
	}

	public boolean isLocked() {
		return locked;
	}

	public void setLocked(boolean locked) {
		this.locked = locked;
	}

	@Override
	public void createUI(GLEx g, int x, int y, LComponent component,
			LTexture[] buttonImage) {

	}

	@Override
	public String getUIName() {
		return "Message";
	}

	@Override
	public void dispose() {
		super.dispose();
		if (print != null) {
			print.dispose();
			print = null;
		}
		if (animation != null) {
			animation.dispose();
			animation = null;
		}
	}
}
